packages/lib/contracts/application/Package.sol

```diff
@@ -11,55 +11,98 @@ import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+    require(!hasVersion(semanticVersion), "Given version is already registered in package");
+    require(!semanticVersionIsZero(semanticVersion), "Version must be non zero");
+
+    // Register version
+    bytes32 versionId = semanticVersionHash(semanticVersion);
+    versions[versionId] = Version(semanticVersion, contractAddress, contentURI);
+    
+    // Update latest major
+    uint64 major = semanticVersion[0];
+    if (major > latestMajor) {
+      latestMajor = semanticVersion[0];
+    }
+
+    // Update latest version for this major
+    uint64 minor = semanticVersion[1];
+    uint64 patch = semanticVersion[2];
+    uint64[3] latestVersionForMajor = versions[majorToLatestVersion[major]].semanticVersion;
+    if (semanticVersionIsZero(latestVersionForMajor) // No latest was set for this major
+       || (minor > latestVersionForMajor[1]) // Or current minor is greater 
+       || (minor == latestVersionForMajor[1] && patch > latestVersionForMajor[2]) // Or current patch is greater
+       ) { 
+      majorToLatestVersion[major] = versionId;
+    }
+
+    emit VersionAdded(semanticVersion, contractAddress, contentURI);
   }
 
   /**
    * @dev Checks whether a version is present in the package.
-   * @param version Name of the version.
+   * @param semanticVersion Name of the version.
    * @return true if the version is already in the package, false otherwise.
    */
-  function hasVersion(string version) public view returns (bool) {
-    return address(versions[version]) != address(0);
+  function hasVersion(uint64[3] semanticVersion) public view returns (bool) {
+    Version storage version = versions[semanticVersionHash(semanticVersion)];
+    return address(version.contractAddress) != address(0);
   }
 
-  /**
-   * @dev Returns the implementation address for a given version and contract name.
-   * @param version Name of the version.
-   * @param contractName Name of the contract.
-   * @return Address where the contract is implemented.
-   */
-  function getImplementation(string version, string contractName) public view returns (address) {
-    ImplementationProvider provider = getVersion(version);
-    return provider.getImplementation(contractName);
+  function getLatest() public view returns (uint64[3] semanticVersion, address contractAddress, bytes contentURI) {
+    return getLatestByMajor(latestMajor);
+  }
+
+  function getLatestByMajor(uint64 major) public view returns (uint64[3] semanticVersion, address contractAddress, bytes contentURI) {
+    Version storage version = versions[majorToLatestVersion[major]];
+    return (version.semanticVersion, version.contractAddress, version.contentURI); 
+  }
+
+  function semanticVersionHash(uint64[3] version) internal pure returns (bytes32) {
+    return keccak256(abi.encodePacked(version[0], version[1], version[2]));
+  }
+
+  function semanticVersionIsZero(uint64[3] version) internal pure returns (bool) {
+    return version[0] == 0 && version[1] == 0 && version[2] == 0;
```

Comment:
I know that we work towards simplifying the architecture of contracts, but it may make sense having the `interface` in a separate file to be clearer